xclim/core/calendar.py

```diff
@@ -20,21 +24,289 @@
+    ----------
+    source: Union[xr.DataArray, xr.Dataset]
+      The source data to interpolate, must have a time coordinate of a valid dtype (np.datetime64 or cftime objects)
+    target: xr.DataArray
+      The target time coordinate of a valid dtype (np.datetime64 or cftime objects)
+
+    Return
+    ------
+    Union[xr.DataArray, xr.Dataset]
+      The source interpolated on the decimal years of target,
+    """
+    cal_src = get_calendar(source)
+    cal_tgt = get_calendar(target)
+
+    out = source.copy()
+    out["time"] = datetime_to_decimal_year(source.time, calendar=cal_src).drop_vars(
+        "time"
+    )
+    target_idx = datetime_to_decimal_year(target, calendar=cal_tgt)
+    out = out.interp(time=target_idx)
+    out["time"] = target
+    return out
+
+
+def _convert_datetime(
+    datetime: Union[pydt.datetime, cftime.datetime],
+    new_doy: Optional[Union[float, int]] = None,
+    calendar: str = "standard",
+):
+    """Convert a datetime object to another calendar.
+
+    Nanosecond information are lost as cftime.datetime doesn't support them.
+
+    Parameters
+    ----------
+    datetime: Union[datetime.datetime, cftime.datetime]
+      A datetime object to convert.
+    new_doy:  Optional[Union[float, int]]
+      Allows for redefening the day of year (thus ignoring month and day information from the source datetime).
+    calendar: str
+      The target calendar
+
+    Returns
+    -------
+    Union[cftime.datetime, pydt.datetime, np.nan]
+      A datetime object of the target calendar with the same year, month, day and time as the source (month and day according to `new_doy` if given).
+      If the month and day doesn't exist in the target calendar, returns np.nan. (Ex. 02-29 in "noleap")
+    """
+    if new_doy is not None:
+        new_date = cftime.num2date(
+            new_doy - 1, f"days since {datetime.year}-01-01", calendar=calendar
+        )
+    else:
+        new_date = datetime
+    try:
+        return datetime_classes[calendar](
+            datetime.year,
+            new_date.month,
+            new_date.day,
+            datetime.hour,
+            datetime.minute,
+            datetime.second,
+            datetime.microsecond,
+        )
+    except ValueError:
+        return np.nan
+
+
+def ensure_cftime_array(time):
+    """Convert an input 1D array to an array of cftime objects. Python's datetime are converted to cftime.DatetimeGregorian.
+
+    Raises ValueError when unable to cast the input.
+    """
+    if isinstance(time, xr.DataArray):
+        time = time.indexes["time"]
+    elif isinstance(time, np.ndarray):
+        time = pd.DatetimeIndex(time)
+    if isinstance(time[0], cftime.datetime):
+        return time
+    if isinstance(time[0], pydt.datetime):
+        return np.array(
+            [cftime.DatetimeProlepticGregorian(*ele.timetuple()[:6]) for ele in time]
+        )
+    raise ValueError("Unable to cast array to cftime dtype")
+
+
+def datetime_to_decimal_year(
```

Comment:
No because that uses nanoseconds since 1st jan 1970. Meaning, that, for example, there would be a 13 days difference in the index of 2020-04-27 (standard) and 2020-04-27 (no leap). I want the midnights of  the 1st of Januaries to align exactly.